React/CSS-in-JS/Chakra.md

```diff
@@ -0,0 +1,657 @@
+For Chakra UI to work correctly, you need to set up the `ChakraProvider` at the root of your application.
+Under the hood Chakra UI is using [emotion ThemeProvider](https://github.com/chakra-ui/chakra-ui/blob/develop/packages/system/src/providers.tsx)
+
+Example:
+
+```jsx
+import { ChakraProvider } from "@chakra-ui/react";
+
+function App({ Component, pageProps }: AppProps): ReactElement {
+  return (
+    <ChakraProvider theme={theme}>
+      <Component {...pageProps} />
+    </ChakraProvider>
+  );
+}
+
+export default App;
+```
+
+Note:
+
+- For Next.js, you need to set this up in `pages/_app.tsx`
+- For Create React App, you need to set this up in `index.tsx`
+
+For including `<CSSReset />` just add `resetCSS` prop to `<ChakraProvider>`
+
+[All provider props](https://chakra-ui.com/docs/getting-started#chakraprovider-props)
+
+### Style props
+
+Style props are a way to alter the style of a component by simply passing props to it. It helps to save time by providing helpful shorthand ways to style components.
+Alternative to that is using chakra factory.
+
+Style props should be used for one-off styles (up to max 3 props):
+
+```tsx
+<Button display="block" mt="md">
+  Click me!
+</Button>
+```
+
+Components that have more than 3 styles should be moved to separate component.
+Every `Box` or `Flex` components should be moved to separate component, regarding of number of styles, with a descriptive name.
+
+(NOTE: Where to create that component will be added after we decide!)
+
+```tsx
+const ListWrapper = chakra(Box, {
+  baseStyle: {
+    shadow: "lg",
+    rounded: "lg",
+    bg: "white",
+  },
+});
+```
+
+For more about style props read [docs](https://chakra-ui.com/docs/features/style-props)
+
+`sx` prop works similar to `style` prop. It takes an object with styles.
+
+Example:
+
+```tsx
+<Button
+  sx={{
+    display: "block",
+    mt: "md",
+    color: "primary.100",
+    mx: 12,
+  }}
+>
+  Click me!
+</Button>
+```
+
+There is also `__css` prop that is similar to `sx`. The only difference is that `__css` prop will be merged before `sx` prop ([github](https://github.com/chakra-ui/chakra-ui/blob/085891be806b855af90b86367f5b26e8151c3ff5/packages/system/src/system.ts#L106)).
+
+### Chakra Factory
+
+Chakra factory serves as an object of chakra JSX elements, and also a function that can be used to enable custom component receive chakra's style props.
+
+```tsx
+import { chakra } from "@chakra-ui/react";
+
+// preferred way!
+chakra("button", {
+  baseStyle: {
+    shadow: "lg",
+    rounded: "lg",
+    bg: "white",
+  },
+});
+
+// using sx prop
+<chakra.button
+  sx={{
+    shadow: "lg",
+    rounded: "lg",
+    bg: "white",
+  }}
+>
+  Click me
+</chakra.button>;
+```
+
+This reduces the need to create custom component wrappers and name them. Syntax is available for common html elements. See the reference for the full [list of elements](https://github.com/chakra-ui/chakra-ui/blob/develop/packages/system/src/system.utils.ts#L9) supported.
```

Comment:
The list just contains elements that you can put in as a string.
You can easily pass any element to chakra:

```
chakra((props) => <someelement {...props} />, {...})
```

Or since `className` is how chakra styles components:

```
chakra(({className}) => <someelement {...{className}} />, {...})
```